[FIRRTL] Remove use of FunctionLike helpers in LowerTypes.

mikeurbach opened the pull request:
The previous implementation relied on the setArgAttrs helper to copy
over the name attribute for every new argument added to the
block. Instead, this accumulates an array of attributes, and sets all
of them in one shot at the end.

Besides setArgAttrs, this also removes the eraseArguments helper. The
performance is subpar compared to directly setting all the attributes
at the end. This means we need to manually remove the block arguments,
and directly set the module's type attribute, instead of relying on
the helper.

With these changes, we take over the bookkeeping work done by the
upstream MLIR helpers to improve performance. The upstream helpers
have to be very careful to keep attributes in a consistent state;
by doing it ourselves, we improve performance by leaving the attributes in an
inconsistent state until the end of the pass.

mikeurbach commented:
This is part of #615. Hopefully the last part. Here are some flame graphs again.

Before (same as the After in #643):

![](https://gist.githubusercontent.com/mikeurbach/b93fb4ab4646679e45950050d7eee916/raw/33e9b3de76b7b0b6978207409e4e9d387ede3078/perf-set-types-once.svg)

After:

![](https://gist.githubusercontent.com/mikeurbach/4489da211b40483bf1262ac6b5340563/raw/555ae6e27e28e734fd23a983f188199a59205a7e/perf-without-helpers.svg)

In my machine, LowerTypes no longer takes the vast majority of the time to run the `test1` performance test:

```
===-------------------------------------------------------------------------===                                                                                                                                    
                      ... Pass execution timing report ...                                                                                                                                                         
===-------------------------------------------------------------------------===                                                                                                                                    
  Total Execution Time: 51.8301 seconds                                                                                                                                                                            
                                                                                                                                                                                                                   
   ---User Time---   ---Wall Time---  --- Name ---                                                                                                                                                                 
   1.4921 (  2.0%)     1.4921 (  2.9%)  CSE                                                                                                                                                                        
   0.0285 (  0.0%)     0.0285 (  0.1%)    (A) DominanceInfo                                                                                                                                                        
   8.9811 ( 12.1%)     8.9811 ( 17.3%)  Canonicalizer                                                                                                                                                              
   1.8768 (  2.5%)     1.8768 (  3.6%)  'firrtl.circuit' Pipeline                                                                                                                                                  
   1.8768 (  2.5%)     1.8768 (  3.6%)    'firrtl.module' Pipeline                                                                                                                                                 
   1.8768 (  2.5%)     1.8768 (  3.6%)      LowerFIRRTLTypes                                                                                                                                                       
   1.4306 (  1.9%)     1.4306 (  2.8%)  LowerFIRRTLToRTLModule                                                                                                                                                     
  47.5987 ( 64.2%)    25.2402 ( 48.7%)  'rtl.module' Pipeline                                                                                                                                                      
   8.9430 ( 12.1%)     3.9919 (  7.7%)    LowerFIRRTLToRTL                                                                                                                                                         
  38.6557 ( 52.1%)    21.2483 ( 41.0%)    RTLCleanup                                                                                                                                                               
   2.8924 (  3.9%)     2.8924 (  5.6%)  CSE                                                                                                                                                                        
   0.2214 (  0.3%)     0.2214 (  0.4%)    (A) DominanceInfo                                                                                                                                                        
   9.9168 ( 13.4%)     9.9168 ( 19.1%)  Canonicalizer                                                                                                                                                              
  74.1886 (100.0%)    51.8301 (100.0%)  Total  
```

mikeurbach commented:
> This is mildly fragile, since we are doing a wholesale update of all
attributes, and we need to take care to persist attributes other than
the argument attributes.

Some more commentary about this: I tried a few different ways to get the argument attributes updated (we have tests that will fail if the name attribute is not correctly persisted). None of them were totally satisfactory, and this was the best solution in terms of performance. What I really wanted was a way to say: "give me all the op's attributes that _are not_ argument attributes" up front. 

Now that I'm typing this, I had one other thought: I guess I could stash all the initial attributes into a set, then loop through the arguments and discard each of the arg attributes from that set... any better ideas?

mikeurbach commented:
> I guess I could stash all the initial attributes into a set, then loop through the arguments and discard each of the arg attributes from that set

I took a crack at this in the most recent commit. The flame graph and pass timing are similar:

![](https://gist.githubusercontent.com/mikeurbach/a92a9ef2f36ed1ed286604243f9e7b25/raw/974b1209071455321e0ecd6ed5fc535f0a1ac1bd/perf-without-helpers-2.svg)

```
===-------------------------------------------------------------------------===                                                                                                                                    
                      ... Pass execution timing report ...                                                                                                                                                         
===-------------------------------------------------------------------------===                                                                                                                                    
  Total Execution Time: 52.8020 seconds                                                                                                                                                                            
                                                                                                                                                                                                                   
   ---User Time---   ---Wall Time---  --- Name ---                                                                                                                                                                 
   1.6077 (  2.1%)     1.6077 (  3.0%)  CSE                                                                                                                                                                        
   0.0312 (  0.0%)     0.0312 (  0.1%)    (A) DominanceInfo                                                                                                                                                        
   9.6120 ( 12.3%)     9.6120 ( 18.2%)  Canonicalizer                                                                                                                                                              
   2.1354 (  2.7%)     2.1354 (  4.0%)  'firrtl.circuit' Pipeline                                                                                                                                                  
   2.1354 (  2.7%)     2.1354 (  4.0%)    'firrtl.module' Pipeline                                                                                                                                                 
   2.1354 (  2.7%)     2.1354 (  4.0%)      LowerFIRRTLTypes                                                                                                                                                       
   1.5790 (  2.0%)     1.5790 (  3.0%)  LowerFIRRTLToRTLModule                                                                                                                                                     
  51.4887 ( 65.9%)    26.1220 ( 49.5%)  'rtl.module' Pipeline                                                                                                                                                      
  10.3270 ( 13.2%)     4.6857 (  8.9%)    LowerFIRRTLToRTL                                                                                                                                                         
  41.1617 ( 52.7%)    21.4363 ( 40.6%)    RTLCleanup                                                                                                                                                               
   2.6097 (  3.3%)     2.6097 (  4.9%)  CSE                                                                                                                                                                        
   0.2378 (  0.3%)     0.2378 (  0.5%)    (A) DominanceInfo                                                                                                                                                        
   9.1361 ( 11.7%)     9.1361 ( 17.3%)  Canonicalizer                                                                                                                                                              
  78.1687 (100.0%)    52.8020 (100.0%)  Total
```

With that, I am feeling much more comfortable with this change, and I think it's ready for review.

lattner commented:
Nice, I'll take a look.  Please remind me to fix RTLCleanup after this lands!

lattner commented on lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp:
```diff
@@ -497,11 +517,22 @@ Value FIRRTLTypesLowering::addArg(Type type, unsigned oldArgNumber,
+      .map([&](NamedAttribute attr) {
+        newAttrs.erase(attr);
+        return attr;
+      });
```
The use of map here is clever, but in a bad way ;-) because I had to dig in to see what is going on at the API level to understand the code.  Please deal with the optional explicitly, something like:

auto argAttrs = module->getAttrDictionary().getNamed(getArgAttrName(oldArgNumber));
if (argAttrs.hasValue()) ...

mikeurbach replied:
:+1: sorry, will stick to good old fashioned `if`.

lattner reviewed (commented):
This general approach looks good, but you don't need a DenseSet of NamedAttrs for 'newAttrs'.

Instead, I'd change this to keep track of two things:
1) a `SmallVector<NamedAttribute>` of new attributes to add, replacing the "insert into newAttrs" logic with a pushback.
2) a `SmallVector<unsigned>` of argument numbers to erase, removing the 'map' + newAttrs.erase with a pushback.

When you go to install the new attributes you can first check to see if either of these is non-empty (if not, nothing to do).  If any is non-empty, then you can build the result SmallVector<NamedAttrs> with three steps:

1) sort the array of erased arguments so you can bin search it
2) copy over old attributes, dropping ones that aren't needed.  Note that attributes are stored in sorted-by-name order in an AttrDict.
3) Append the new attributes,  then create the AttrDict.

This should be nice and fast, and gets out of the hashing and other stuff that DenseSet is doing.

mikeurbach commented:
@lattner thanks for the review, I really appreciate it on these low-level performance optimizations. I will implement the approach you described.

mikeurbach commented:
I switched the implementation to use a `SmallVector<NamedAttribute>` instead of `DenseSet<NamedAttribute>`. The implementation is roughly as described above, except the part about `copy over old attributes, dropping ones that aren't needed` works a little differently.

Since we are copying over more than just argument attributes (`sym_name` is the main one), I couldn't see a nice way to loop through all the attributes and consult the sorted `SmallVector<unsigned>` array to see if the attribute was related to an argument that was removed. Unless I'm missing some API for "give me all the attrs other than arg attrs", that would boil down to attempting to parse the argument index out of the attribute name, which is... wrong.

So instead, I constructed a sorted `SmallVector<NamedAttribute>` array of the argument attributes that are removed, and I consult that to determine if we should copy over a given attribute. Note that in order to keep the relative ordering of arguments as we flatten, we always create a new arg and remove the old arg, so this array is just the initial list of argument attributes. Also note that the unconditional creation of new arguments wasn't the performance issue: it was the getting and setting of argument attributes.

In the end, this uses two `SmallVector<NamedAttribute>` arrays: one of the original argument attributes to ignore, and one of all attributes to add. Using these two arrays instead of a `DenseSet<NamedAttribute>` didn't have any noticeable impact on performance on my machine, but I've implemented it, so I'll push the commit.

lattner commented on lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp:
```diff
@@ -111,6 +114,8 @@ struct FIRRTLTypesLowering : public LowerFIRRTLTypesBase<FIRRTLTypesLowering>,
   Value getBundleLowering(Value oldValue, StringRef flatField);
   void getAllBundleLowerings(Value oldValue, SmallVectorImpl<Value> &results);
 
+  StringRef getArgAttrName(unsigned newArgNumber);
```
This can be more self contained and just as efficient if you make this return an MLIR identifier:

```
Identifier FIRRTLTypesLowering::getArgAttrName(unsigned newArgNumber) {
  SmallString<16> argNameBuffer;
  mlir::impl::getArgAttrName(newArgNumber, argNameBuffer);
  return Identifier::get(argNameBuffer, context);
}
```

That will eliminate the argNameBuffer.

lattner commented on lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp:
```diff
@@ -497,11 +531,14 @@ Value FIRRTLTypesLowering::addArg(Type type, unsigned oldArgNumber,
+    auto newArgNumber = newValue.getArgNumber() - originalNumArgs;
+    auto newArgAttrName = getArgAttrName(newArgNumber);
+    auto newAttr = builder->getNamedAttr(newArgAttrName, newArgAttrs);
+    newAttrs.push_back(newAttr);
```
Once getArgAttrName returns an identifier, you can just use:

`newAttrs.push_back(NamedAttribute(newArgAttrName, newArgAttrs));`

or:

`newAttrs.push_back({newArgAttrName, newArgAttrs});`

instead of calling getNamedAttr

lattner commented on lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp:
```diff
@@ -136,8 +146,16 @@ void FIRRTLTypesLowering::runOnOperation() {
+  DictionaryAttr originalAttrs = module->getAttrDictionary();
+  for (size_t i = 0, e = module.getNumArguments(); i < e; ++i)
+    originalArgAttrs.push_back(
+        originalAttrs.getNamed(getArgAttrName(i)).getValue());
```
I think it possible to sink this loop to after the "if argstoremove.empty()" check below, which avoids work in the case where no ports are lowered.

lattner commented on lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp:
```diff
@@ -153,12 +171,28 @@ void FIRRTLTypesLowering::runOnOperation() {
+  // Copy over any attributes that weren't original argument attributes.
+  auto *argAttrBegin = originalArgAttrs.begin();
+  auto *argAttrEnd = originalArgAttrs.end();
+  std::sort(argAttrBegin, argAttrEnd);
```
Can this use llvm::sort?

lattner replied:
Actually, I think this is already always sorted now, because DictionaryAttr always stors its attributes in sorted order (to canonicalize the dictionary).  I think you can even eliminate the originalArgAttrs copy and just use lowerbound over the elements in the dictionary attr itself.

mikeurbach replied:
I wanted to take advantage of the fact that DictionaryAttr is sorted already, but I couldn't figure out how to track just the attributes we want to erase without copying them into some sort of array. The way that I'm doing this with `originalAttrs.getNamed(getArgAttrName(i))` builds that array in numerical order of arguments like `(arg0, arg1, arg2... arg10, arg11... arg20)`, but lower_bound expects them sorted like `(arg0... arg1, arg10, arg11... arg2, arg20)`. The FunctionSupport helper [getAllArgAttrs](https://github.com/llvm/llvm-project/blob/f7e675b3da1d7afd43af51c5a5e9140c87ef945d/mlir/include/mlir/IR/FunctionSupport.h#L314-L317) seems to build up the result in the same order, so I'm not sure if there is any way to track the original argument attributes and keep them sorted by name.

I was able to at least sink the loop that builds the array after the `if argsToRemove.empty()` check, and keep it sorted using the `DictionaryAttr::sortInPlace` helper (which has some optimizations and calls llvm::array_pod_sort in the general case) instead of `std::sort`.

lattner reviewed (commented):
This is looking great Mike!

darthscsi reviewed (commented):
It could go in with lattner's suggestions.

That said, as I looked at it, it seemed like there should be an easier way.  These changes move from one-at-a-time field lowering to one-at-a-time arg adding, but bulk attribute adding.  Would it be easier just to compute the new argument and attribute set on the side then bulk swapping them in?  I'm primarily concerned that body->addArgument in addArg is adding field at a time, but the attributes are deferred.  I don't know if it would be cleaner, but it's a thought.

mikeurbach commented:
Thanks @lattner for taking another look, I will incorporate that feedback.

> Would it be easier just to compute the new argument and attribute set on the side then bulk swapping them in?

@darthscsi I will think about that, but the current approach seems to be the simplest to me (that has good performance).

For attributes, it's easy to bulk swap them in at the end because they're not actually used in the pass, and we just need them to keep the contract of FunctionLike operations (i.e. there is a `sym_name` attr, a `type` attr with FunctionType, and `argN` attrs for each N args).

For the block arguments themselves, we actually need to get create fresh `Value`s that are used throughout the pass, and these need to be created up front. Repeated calls to `body->addArgument` seemed like this simplest way to manifest new `Value`s. In my tests, doing this at the beginning is not the bottleneck, but repeated calls to `setAttr` and `removeAttr` are (you can click into the original flamegraph and see that's where `OpToOpPassAdaptor::run` spent most of its time).

mikeurbach commented:
Here's the latest flamegraph for the same test1.fir test case:

![](https://gist.githubusercontent.com/mikeurbach/475ee55867f986066a326993d99e78c3/raw/9ceba62a62f00a236102d1423ec41002a70263f8/perf-without-helpers-latest.svg)

mikeurbach commented:
@lattner @darthscsi please let me know what you think about the current implementation. I tried to address most of the concerns, but see my comments above for where I didn't. This seems to be the best performing revision yet on [this](https://github.com/circt/perf/blob/trunk/regress/test1.fir) test case.

lattner commented:
Thanks, I'm taking a look.  If I understand the flamegraph correctly it sounds like we need a cache for hasNoMemoryEffect.  I wonder what is calling it that much in lower types?

lattner commented on lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp:
```diff
@@ -131,6 +136,10 @@ struct FIRRTLTypesLowering : public LowerFIRRTLTypesBase<FIRRTLTypesLowering>,
   DenseMap<ValueIdentifier, Value> loweredBundleValues;
+
+  // State to track the new attributes for the module.
+  SmallVector<NamedAttribute, 8> newAttrs;
```
would it make sense to rename this to `newModuleAttrs` and `originalNumModuleArgs` for clarity?

lattner commented on lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp:
```diff
@@ -164,12 +174,35 @@ void FIRRTLTypesLowering::runOnOperation() {
+  auto *argAttrEnd = originalArgAttrs.end();
+  for (auto attr : originalAttrs)
+    if (std::lower_bound(argAttrBegin, argAttrEnd, attr) == argAttrEnd)
+      newAttrs.push_back(attr);
```
This is looking good, but I think it can still be made better.  Feel free to land this and we can keep improving in place though!

This can be fundamentally a single pass algorithm, something like this:

```
for (auto attr : module->getAttrDictionary()) {
   // If this is the name of an argument attribute, return the number for it.
  Optional<unsigned> argNumber = getArgAttrNumber(attr.name);
  if (!argNumber.hasValue() || weAreKeepingThisArgument(argNumber.get())) 
     newAttrs.push_back(attr);
}
```
It requires introducing a "getArgAttrNumber" sort of helper function (which should be generally useful anyway, akin to `getArgAttrName`) and a way to know which argument ID's are to be deleted (which could be stored in a SmallSet next to argsToRemove.

mikeurbach replied:
Yep, I have had that thought, but wasn't sure if defining `getArgAttrNumber(Identifier)` was going to be too hacky/stringly-typed (see comment [here](https://github.com/llvm/circt/pull/651#issuecomment-786296652)).

lattner replied:
Thanks.  I think the arg name attribute design is inherently stringly typed, so embracing that with an "atoi" makes sense.   StringRef has a helper for this.

mikeurbach commented:
Thanks for the iterations @lattner. Just landed this and we can continue iterating from here.
